src/ha-utils/ha/common/utils/apiutils/RESTUtil.go

```diff
@@ -0,0 +1,126 @@
+package apiutils
+
+import (
+	"bytes"
+	"encoding/json"
+	"ha-utils/ha/common/constants"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+)
+
+func InvokeRESTAPI(httpMethod string, apiURL string, reqHeader map[string]string, requestBody interface{}) (string, string, bool) {
+
+	var httpClient *http.Client = &http.Client{}
+	var request *http.Request
+	var response *http.Response
+	var responseStatus string
+	var err error
+	var responseStatusSuccessful bool
+
+	log.Println("URL being hit is ", httpMethod, " ", apiURL)
+
+	if requestBody == nil {
+		request, err = http.NewRequest(httpMethod, apiURL, nil)
+	} else {
+		jsonValue, _ := json.Marshal(requestBody)
+		request, err = http.NewRequest(httpMethod, apiURL, bytes.NewBuffer(jsonValue))
+	}
+	if err != nil {
+		log.Println("Error occured while creating new request")
+		return "Error occured while creating new HTTP Request", "", false
+	}
+
+	// TODO: set appropriate content type (may be application/json) if not set
+	// Include all required headers passed by the caller
+	for key, value := range reqHeader {
+		request.Header.Set(key, value)
+	}
+
+	// All required params have been set - call the api.
+	response, err = httpClient.Do(request)
+	if err != nil {
+		log.Println("Error occured while caling the api: ", apiURL, "Error is ", err.Error())
+		return err.Error(), "", false
+	}
+
+	responseStatus = strings.TrimSpace(strings.ToUpper(response.Status))
+	log.Println("HTTP Response Status Code: ", responseStatus)
+	// Check if the response is successful by mathcing for 2xx
+	responseStatusSuccessful, _ = regexp.MatchString("^2[0-9][0-9] ", responseStatus)
+	if !responseStatusSuccessful {
+		log.Println("API call was not successful : ", apiURL)
+	}
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		log.Println("HTTP Status : ", responseStatus, "Error occurred while reading response body ", err.Error())
+		return err.Error(), "", false
+	}
+	//	log.Println("response Body:", string(body))
+	defer response.Body.Close()
+
+	if responseStatusSuccessful {
+		return string(body), responseStatus, true
+	} else {
+		log.Println("Response body:", string(body))
```

Comment:
will code ever enter this block? wouldnt it execute 59,60 during error flows? seems too many branching for success/failure flows. Ideally should have just two.

Reply:
This checks for the status code of the response. If the response is not having '2**' status code (* being a digit from 0-9), `responseStatusSuccessful` is false.
This check is for that condition. 
In lines 59, 60; it is checking for if the reading of the body of response was successful/unsuccessful.

So, for ex, the response is 404, and the response body was read fine, then this block's decision will be executed.

Agreed. The flow could have been written better.